source/server/listener_manager_impl.cc

```diff
@@ -580,34 +670,77 @@ const Network::FilterChain*
 ListenerImpl::findFilterChainForSourceTypes(const SourceTypesArray& source_types,
                                             const Network::ConnectionSocket& socket) const {
 
-  auto filter_chain_local =
+  const auto& filter_chain_local =
       source_types[envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType::
                        FilterChainMatch_ConnectionSourceType_LOCAL];
 
-  auto filter_chain_external =
+  const auto& filter_chain_external =
       source_types[envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType::
                        FilterChainMatch_ConnectionSourceType_EXTERNAL];
 
-  // isLocalConnection can be expensive. Call it only if LOCAL or EXTERNAL are defined.
-  const bool is_local_connection = (filter_chain_local || filter_chain_external)
-                                       ? Network::Utility::isLocalConnection(socket)
-                                       : false;
+  // isLocalConnection can be expensive. Call it only if LOCAL or EXTERNAL have entries.
+  const bool is_local_connection =
+      (!filter_chain_local.first.empty() || !filter_chain_external.first.empty())
+          ? Network::Utility::isLocalConnection(socket)
+          : false;
 
   if (is_local_connection) {
-    if (filter_chain_local) {
-      return filter_chain_local.get();
+    if (!filter_chain_local.first.empty()) {
+      return findFilterChainForSourceIpAndPort(*filter_chain_local.second, socket);
     }
   } else {
-    if (filter_chain_external) {
-      return filter_chain_external.get();
+    if (!filter_chain_external.first.empty()) {
+      return findFilterChainForSourceIpAndPort(*filter_chain_external.second, socket);
     }
   }
 
-  auto filter_chain_any =
+  const auto& filter_chain_any =
       source_types[envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType::
                        FilterChainMatch_ConnectionSourceType_ANY];
 
-  return filter_chain_any.get();
+  if (!filter_chain_any.first.empty()) {
+    return findFilterChainForSourceIpAndPort(*filter_chain_any.second, socket);
+  } else {
+    return nullptr;
+  }
+}
+
+const Network::FilterChain*
+ListenerImpl::findFilterChainForSourceIpAndPort(const SourceIPsTrie& source_ips_trie,
+                                                const Network::ConnectionSocket& socket) const {
+  // Use invalid IP address (matching only filter chains without IP requirements) for UDS.
+  static const auto& fake_address = Network::Utility::parseInternetAddress("255.255.255.255");
```

Comment:
This is harder to read than it probably needs to be, with the mix of `static`, ref and const shared ptr.

Reply:
This was copied from the existing code, will fix.